jetty-server/src/main/java/org/eclipse/jetty/server/DetectorConnectionFactory.java

```diff
@@ -0,0 +1,322 @@
+            endPoint.close();
+            throw new IllegalStateException("Cannot upgrade: " + nextConnection + " does not implement " + Connection.UpgradeTo.class.getName());
+        }
+
+        endPoint.upgrade(nextConnection);
+        connector.getByteBufferPool().release(buffer);
+    }
+
+    @Override
+    public Connection newConnection(Connector connector, EndPoint endPoint)
+    {
+        return configure(new DetectorConnection(endPoint, connector, _detectingConnectionFactories, _detectionUnsuccessful), connector, endPoint);
+    }
+
+    class DetectorConnection extends AbstractConnection implements Connection.UpgradeFrom, Connection.UpgradeTo
+    {
+        private final Connector _connector;
+        private final List<Detecting> _detectingConnectionFactories;
+        private final DetectionUnsuccessfulTrampoline _detectionUnsuccessful;
+        private final ByteBuffer _buffer;
+
+        protected DetectorConnection(EndPoint endp, Connector connector, List<Detecting> detectingConnectionFactories, DetectionUnsuccessfulTrampoline detectionUnsuccessful)
+        {
+            super(endp, connector.getExecutor());
+            _connector = connector;
+            _detectingConnectionFactories = detectingConnectionFactories;
+            _detectionUnsuccessful = detectionUnsuccessful;
+            _buffer = connector.getByteBufferPool().acquire(getInputBufferSize(), true);
+        }
+
+        @Override
+        public void onOpen()
+        {
+            super.onOpen();
+
+            if (!detectAndUpgrade())
+                fillInterested();
+        }
+
+        @Override
+        public void onFillable()
+        {
+            try
+            {
+                while (BufferUtil.space(_buffer) > 0)
+                {
+                    // Read data
+                    int fill = getEndPoint().fill(_buffer);
+                    if (fill < 0)
+                    {
+                        getEndPoint().shutdownOutput();
+                        return;
+                    }
+                    if (fill == 0)
+                    {
+                        fillInterested();
+                        return;
+                    }
+
+                    if (detectAndUpgrade())
+                        return;
+                }
+
+                // all Detecting instances want more bytes than this buffer can store
+                LOG.warn("Failed to detect upgrade target on {} for {}", _detectingConnectionFactories, getEndPoint());
+                releaseAndClose();
+            }
+            catch (Throwable x)
+            {
+                LOG.warn("DetectorConnection error for " + getEndPoint(), x);
+                releaseAndClose();
+            }
+        }
+
+        /**
+         * @return true when upgrade was performed, false otherwise.
+         */
+        private boolean detectAndUpgrade()
+        {
+            boolean noMagicFound = true;
```

Comment:
Drop "magic", you can just use `boolean detected = false`.

Reply:
Using `boolean detected = false` would inverse the logic so I renamed the variable `detectionUnsuccessful` instead.